### ISSUE TITLE
"UnreadableSessionDataException Unreadable session ..." after upgrading to 9.4.27

### ISSUE THREAD

**Issue by hanswesterbeek** [eclipse/jetty.project#4682]
**Jetty version** 9.4.27.v20200227

**Java version** openjdk11

**OS type/version** centos

**Description**

After upgrading from 9.4.25.v20191220 to the above, we often see Warnings like this in our logs: 
`UnreadableSessionDataException
Unreadable session node0q5obwgxy5xig1pk5hkb6lpdbp7 for node0__controlroom_0.0.0.0`

Stacktrace below shows a regular request, but this also occurs when the `Housekeeper` attempts to read session data. The error looks like the `Inputstream` that is passed to the constructor of `ClassLoadingObjectInputStream` is null.

However, I have verified that the table and column do actually contain data:

```sql
select length(map) from jettysessions ;
 length 
--------
   5370
   2804
   4791
   ....
(25 rows)
```

I have dropped and let jetty recreate the table.

Postgres-version is **PostgreSQL 9.3.25**
Jdbc-driver version: **42.2.10** 

Full stacktrace follows:
```
org.eclipse.jetty.server.session.UnreadableSessionDataException: Unreadable session node0q5obwgxy5xig1pk5hkb6lpdbp7 for node0__controlroom_0.0.0.0
    at org.eclipse.jetty.server.session.JDBCSessionDataStore.doLoad(JDBCSessionDataStore.java:640)
    at org.eclipse.jetty.server.session.AbstractSessionDataStore.lambda$load$0(AbstractSessionDataStore.java:93)
    at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1367)
    at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1404)
    at org.eclipse.jetty.server.session.SessionContext.run(SessionContext.java:92)
    at org.eclipse.jetty.server.session.AbstractSessionDataStore.load(AbstractSessionDataStore.java:101)
    at org.eclipse.jetty.server.session.AbstractSessionCache.loadSession(AbstractSessionCache.java:416)
    at org.eclipse.jetty.server.session.AbstractSessionCache.lambda$getAndEnter$0(AbstractSessionCache.java:353)
    at org.eclipse.jetty.server.session.DefaultSessionCache.lambda$doComputeIfAbsent$0(DefaultSessionCache.java:116)
    at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1737)
    at org.eclipse.jetty.server.session.DefaultSessionCache.doComputeIfAbsent(DefaultSessionCache.java:114)
    at org.eclipse.jetty.server.session.AbstractSessionCache.getAndEnter(AbstractSessionCache.java:346)
    at org.eclipse.jetty.server.session.AbstractSessionCache.get(AbstractSessionCache.java:327)
    at org.eclipse.jetty.server.session.SessionHandler.getSession(SessionHandler.java:906)
    at org.eclipse.jetty.server.session.SessionHandler.getHttpSession(SessionHandler.java:552)
    at org.eclipse.jetty.server.session.SessionHandler.checkRequestedSessionId(SessionHandler.java:1656)
    at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1529)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1215)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:717)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
    at org.eclipse.jetty.server.Server.handle(Server.java:500)
    at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
    at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
    at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
    at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
    at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
    at java.lang.Thread.run(Thread.java:834)
java.io.EOFException: null
    at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2763)
    at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3258)
    at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:873)
    at java.io.ObjectInputStream.<init>(ObjectInputStream.java:350)
    at org.eclipse.jetty.util.ClassLoadingObjectInputStream.<init>(ClassLoadingObjectInputStream.java:50)
    at org.eclipse.jetty.server.session.JDBCSessionDataStore.doLoad(JDBCSessionDataStore.java:634)
    at org.eclipse.jetty.server.session.AbstractSessionDataStore.lambda$load$0(AbstractSessionDataStore.java:93)
    at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1367)
    at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1404)
    at org.eclipse.jetty.server.session.SessionContext.run(SessionContext.java:92)
    at org.eclipse.jetty.server.session.AbstractSessionDataStore.load(AbstractSessionDataStore.java:101)
    at org.eclipse.jetty.server.session.AbstractSessionCache.loadSession(AbstractSessionCache.java:416)
    at org.eclipse.jetty.server.session.AbstractSessionCache.lambda$getAndEnter$0(AbstractSessionCache.java:353)
    at org.eclipse.jetty.server.session.DefaultSessionCache.lambda$doComputeIfAbsent$0(DefaultSessionCache.java:116)
    at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1737)
    at org.eclipse.jetty.server.session.DefaultSessionCache.doComputeIfAbsent(DefaultSessionCache.java:114)
    at org.eclipse.jetty.server.session.AbstractSessionCache.getAndEnter(AbstractSessionCache.java:346)
    at org.eclipse.jetty.server.session.AbstractSessionCache.get(AbstractSessionCache.java:327)
    at org.eclipse.jetty.server.session.SessionHandler.getSession(SessionHandler.java:906)
    at org.eclipse.jetty.server.session.SessionHandler.getHttpSession(SessionHandler.java:552)
    at org.eclipse.jetty.server.session.SessionHandler.checkRequestedSessionId(SessionHandler.java:1656)
    at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1529)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1215)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:717)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
    at org.eclipse.jetty.server.Server.handle(Server.java:500)
    at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
    at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
    at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
    at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
    at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
    at java.lang.Thread.run(Thread.java:834)
```

**Comment by hanswesterbeek** [comment]
Possibly related to #4528

**Comment by janbartel** [comment]
I think you're just now seeing the exception that was hidden before by bug #4528 
Seems like the input stream of the attribute blob from the db is being truncated such that the jvm can't even read the first byte to check that it's a serialized object.

A few things:

* turn on DEBUG level logging for org.eclipse.jetty.server.session and repost the log output
* post any configuration settings for your sessions that are not defaults
* have you checked the db logs for any errors?
* do you have more than one node, and if so: a) do you have a sticky load balancer  b) has the session moved node at all?
* can you post the row from the db matching one of the failed sessions?
* maybe post the serialized bytes from the map column for a failed session to me and I'll see if I can analyse it

**Comment by hanswesterbeek** [comment]
First I'll answer the questions, then move onto getting more data (eg logs).

Yes we have sticky sessions on two nodes, but that is not my first suspect.

Upon closer inspection I found out that the codeI do suspect some of the code that sets up the configuration (the code I inherited does this programatically). 

Thank you for your quick reply. So I know believe that this is caused by an imperfect migration from jetty 9.2.x to 9.4.x.

Please close this issue.

**Comment by hanswesterbeek** [comment]
I've just found a session object with empty session data.

```
sql

cr-jetty-session-db=> select length(map) from jettysessions ;
 length 
--------
   2203
   2159
   2804
   2177
   1382
   2220
   6206
   5825
   5807
   2149
   2162
   6921
   6326
   6024
   2664
   2145
   2166
   5508
   2185
   6117
   5496
   2297
   2605
   5332
   2183
   2191
   2070
   2174
   2195
   2192
   2204
   2172
   5296
      0
   6183
   1485
   6344
   5322
   1391
   2223
   2171
   4956
(42 rows)
````

That would indeed lead to the exception above.

What can cause this situation to occur?

**Comment by hanswesterbeek** [comment]
I am also seeing occurances of #4520 in my logs

**Comment by janbartel** [comment]
@hanswesterbeek  a 0 length for the attributes map is not necessarily a problem. If the map contains no attributes, then we write an array of length 0 to the db for that column. It's only a problem if you're expecting there to be attributes.

You'll have to enable DEBUG logging for sessions, and trace through the lifecycle of these sessions, and correlate them with the content of the database.

**Comment by dhawal96** [comment]
@hanswesterbeek @janbartel +1 to this issue as I am also seeing the same java.io.EOFException exception after testing Jetty version 9.4.27.v20200227 and I am also seeing empty session data in the map column. Note that if I downgrade to a lower Jetty version, re-running the same steps for my application does not show empty session data in the serialized bytes from the map column.

**Comment by janbartel** [comment]
@dhawal96 let me repeat that since 9.4.21 you wouldn't have seen any exceptions when trying to write the session, but the failed write would result in the removal of the row, so everything would seem normal.

Can both you and @hanswesterbeek check your database logs and see if there is anything happening at the time of the session write (that could cause the 0 length) and at the time of the session read.

**Comment by janbartel** [comment]
Ah! I think I see a problem: we changed the handling of how we write out sessions to include classloading information in #2932 . As part of that change, we write out 0 bytes if there are no session attributes, but when we need to read the the session attributes, the classloading handling is unable to read the zero bytes as an empty object. I'm working on a fix.  Thanks for persisting with this issue.

**Comment by janbartel** [comment]
@hanswesterbeek and @dhawal96  if either of you have a test bed, please checkout the fix associated with PR #4688, build and verify it's good.

**Comment by hanswesterbeek** [comment]
@janbartel That's great. I don't have a non-risky way of verifying this but I think your diagnosis and solution are pretty conclusive.

Looking forward to .28.
